drivers/d3d12/rendering_device_driver_d3d12.cpp

```diff
@@ -5509,4 +5526,8 @@ RenderingDeviceDriverD3D12::~RenderingDeviceDriverD3D12() {
 	}
 
 	glsl_type_singleton_decref();
+
+	DEV_ASSERT(pools_command_buffers.is_empty());
+
+	resources_allocator.reset(true);
```

Comment:
The drivers are written with the mindset of being as thin wrappers as possible over the underlying API. Therefore, they are not meant to perform the cleanup higher-level classes (i.e., `RenderingDevice`) failed to do. The `RenderingDeviceDriver` should be sure it doesn't leak anything if the higher layer has done proper cleanup. Otherwise, it's OK that validation errors happen. Moreover, adding explicit cleanup at this level would take us back again to re-implementing across the multiple drivers something that could be implemented once at the API-agnostic level.

That mindset involves not doing any extra book-keeping only needed for cleanup, like using `pools_command_buffers` for extra book-keeping. If fact, it's a left-behind of one of my iterations before I decided to couple command list allocator with command list.

Reply:
If that future comes, we can always re-assess. For the time being, I'd advise against breaking the separation of concerns between `RenderingDevice` and the drivers. Are we leaking any command list allocator at the moment? Are the leaked resources also leaked in the Vulkan driver? If no, there's something to be fixed here indeed, but a different thing. If yes, `RenderingDevice` is the one missing cleanup.

Reply:
Yes, I agree with Pedro here, a lot of the responsibilities over cleanup should be moved to RenderingDevice whenever it's possible. Driver should only deallocate whatever hasn't been explicitly allocated by the caller of the driver.